Add Optional JIT compilation

richardpringle opened the pull request:
Closes https://github.com/ithacaxyz/odyssey/issues/25

Will add more of a description later

Commit:
WIP

richardpringle committed:
WIP

richardpringle committed:
WIP

richardpringle committed:
WIP

richardpringle commented on crates/node/src/compiler.rs:
```diff
@@ -0,0 +1,147 @@
+            let cache = cache.clone();
+
+            move || {
+                let context = Box::leak(Box::new(revmc::llvm::Context::create()));
```
Can I leak this, does that make sense?  I guess I shouldn't need to since `context` will live for the whole `while` loop below.

DaniPopes replied:
You should be able to use `with_llvm_context` and nest the implementation inside of the closure

richardpringle committed:
Don't mess around with the interpreter's memory

richardpringle commented on crates/node/src/compiler.rs:
```diff
@@ -0,0 +1,144 @@
+                    dbg!("compiled", &name);
+
+                    let result =
+                        unsafe { compiler.jit(&name, &code, spec_id) }.expect("catastrophe");
```
I'm a little confused as to the relationship between a _compiler_, a _module_, and a _contract_. 
Should I be compiling all the code with the same compiler?

DaniPopes commented on crates/node/src/compiler.rs:
```diff
@@ -0,0 +1,144 @@
+
+                    cache.lock().unwrap().insert(hash, result);
+
+                    unsafe { compiler.clear().expect("could not clear") };
```
> Should I just be creating a new compiler every time and storing the compiler as well?

Yes that's right

DaniPopes commented on crates/node/src/evm.rs:
```diff
@@ -247,7 +251,14 @@ impl ConfigureEvm for OdysseyEvmConfig {
+        // TODO: not sure if this is correct, I don't see a call-site for this method
+        let spec_id = revm_spec(chain_spec, &Head::default());
+
+        compiler::ExternalContext::new(spec_id)
```
spec ID should also be dynamic and received from the channel

richardpringle committed:
Create a new compiler on every message received

richardpringle committed:
Send spec-id with each compilation

richardpringle committed:
Move the compiler to the Odyssey config

I thought that the context was created once per config, but that was not the case. This commit moves the compiler thread initialization to the config initialization and it all seems to work.

richardpringle committed:
Move the compiling logic back to the compiler module

richardpringle committed:
Add comment
